allennlp/predictors/predictor.py

```diff
@@ -57,17 +63,157 @@ def predict_json(self, inputs: JsonDict) -> JsonDict:
+        return new_instances
+
+    def get_gradients(self, instances: List[Instance]) -> Dict[str, np.ndarray]:
+        """
+        Gets the gradients of the loss with respect to the model inputs. 
+
+        Parameters
+        ----------
+        instances: List[Instance]
+
+        Returns
+        -------
+        Dict[str, np.ndarray]
+            Dictionary of gradient entries for each input fed into the model.
+            The keys have the form ``{grad_input_1: ..., grad_input_2: ... }``
+            up to the number of inputs given. 
+            
+        Notes
+        -----
+        Takes a ``JsonDict`` representing the inputs of the model and converts
+        them to :class:`~allennlp.data.instance.Instance`s, sends these through
+        the model :func:`forward` function after registering hooks on the embedding
+        layer of the model. Calls :func:`backward` on the loss and then removes the
+        hooks. 
+        """
+        self._register_hooks()
+
+        dataset = Batch(instances)
+        dataset.index_instances(self._model.vocab)
+ 
+        outputs = self._model.decode(self._model.forward(**dataset.as_tensor_dict()))
+
+        loss = outputs['loss']
+
+        self._model.zero_grad()
+
+        loss.backward()
+
+        # Remove hooks 
+        for hook in self.hooks:
+            hook.remove()
+
+        grad_dict = dict()
+        for idx, grad in enumerate(self.extracted_grads):
+            key = 'grad_input_' + str(idx + 1)
+            # Squeeze to remove batch dimension
+            grad_dict[key] = grad.squeeze_(0).detach().cpu().numpy()
+
+        return grad_dict, outputs 
+
+    def _register_hooks(self):
+        """
+        Registers a backward hook on the 
+        :class:`~allennlp.modules.text_field_embedder.basic_text_field_embbedder.BasicTextFieldEmbedder`
+        class. 
+        """
+        # For multiple inputs, the hook will be called multiple times
+        # so we append the incoming gradients to a list
+        self.extracted_grads = []
+        self.hooks = []
+
+        def hook_layers(module, grad_in, grad_out):
+            self.extracted_grads.append(grad_out[0])
+
+        # Register the hooks
+        for module in self._model.modules():
+            if isinstance(module, BasicTextFieldEmbedder):
+                backward_hook = module.register_backward_hook(hook_layers)
+                self.hooks.append(backward_hook)
+
+    def _normalize(self, grads: Dict[str, np.ndarray]) -> Dict[str, np.ndarray]:
+        """
+        Normalize the gradients into the range [0,1] for visualization. 
+        Notes
+        -----
+        0.0 is dark red and 1.0 is dark blue. 0.5 is no highlight. 
+        This normalization step is very rarely discussed in the 
+        literature and the interpretation is actually quite sensitive to it. 
+        In the code below, we normalize scores across the words, doing positive 
```

Comment:
So, if I have scores of [-100, -70, 0, 0.1, 0.2], this will say that -100 is just as "negative" as 0.2 is "positive"?  Why is this the right thing to do?